extensions/libevent.c

```diff
@@ -0,0 +1,318 @@
+}
+
+static void
+cb_func (evutil_socket_t fd, short what, void *arg)
+{
+  struct wait_data* data = arg;
+  int rv = data->rv;
+
+  // In theory, we should never exceed the maximum number of events. This is
+  // because every time we add a new FD to monitor we check if we have enough
+  // room in our vector and if not we resize it. So, if this condition is true
+  // is because those assumptions are wrong and we might be doing something
+  // funky.
+  if (rv >= data->maxevents)
+    {
+      scm_puts ("fibers-libevent[ERROR]: ", scm_current_error_port ());
+      scm_puts ("max events fired on [", scm_current_error_port ());
+      scm_display (scm_from_int(fd), scm_current_error_port ());
+      scm_puts ("], ignoring.\n", scm_current_error_port ());
+      return;
+    }
+
+  struct event_data ev_data = { fd, what };
+  memcpy (data->events + rv, &ev_data, sizeof (struct event_data));
+
+  data->rv += 1;
+}
+
+static SCM
+scm_primitive_event_wake (SCM wakefd)
+#define FUNC_NAME "primitive-event-wake"
+{
+  int c_fd;
+  char zero = 0;
+
+  c_fd = scm_to_int (wakefd);
+
+  if (write (c_fd, &zero, 1) <= 0 && errno != EWOULDBLOCK && errno != EAGAIN)
+    SCM_SYSERROR;
+
+  return SCM_UNSPECIFIED;
+}
+#undef FUNC_NAME
+
+static SCM
+scm_primitive_create_event_base (SCM eventsv)
+#define FUNC_NAME "primitive-create-event-base"
+{
+  struct libevt_data *libevt;
+  struct wait_data *data;
+  struct event_base *base;
+
+  if ((base = event_base_new ()) == NULL)
+    scm_syserror ("couldn't instantiate event_base");
+
+  libevt = (struct libevt_data *) scm_gc_malloc (sizeof (struct libevt_data),
+                                                 "libevt_data");
+  libevt->base = base;
+
+  data = (struct wait_data *) scm_gc_malloc (sizeof (struct wait_data),
+                                             "wait_data");
+  data->rv = 0;
+  data->events = (struct event_data *) SCM_BYTEVECTOR_CONTENTS (eventsv);
+  data->maxevents = SCM_BYTEVECTOR_LENGTH (eventsv) / sizeof (struct event_data);
+
+  return scm_list_2 (scm_from_pointer (libevt, free_libevt),
+                     scm_from_pointer (data, free_wait_data));
+}
+#undef FUNC_NAME
+
+static SCM
+scm_primitive_add_event (SCM lst, SCM fd, SCM ev)
+#define FUNC_NAME "primitive-add-event"
+{
+  int c_fd;
+  short c_ev;
+  struct event *event;
+  struct wait_data *data;
+  struct libevt_data *libevt;
+
+  c_fd = scm_to_int (fd);
+  c_ev = scm_to_short (ev);
+
+  libevt =
+    (struct libevt_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (0)));
+  data =
+    (struct wait_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (1)));
+
+  event = event_new (libevt->base, c_fd, c_ev, cb_func, data);
+  event_add (event, NULL);
+
+  return scm_from_pointer (event, NULL);
+}
+#undef FUNC_NAME
+
+static SCM
+scm_primitive_remove_event (SCM lst, SCM scm_event)
+#define FUNC_NAME "primitive-remove-event"
+{
+  struct event *event = scm_to_pointer(scm_event);
+
+  event_del (event);
```

Comment:
According to the documentation, somehow this can error out, so exceptions!

Reply:
done.